### PR TITLE
RAM minter

### PR DESCRIPTION
**Pull request by ryley-o**
## Description of the change

Add a new RAM minter.

Please see internal design document for minter implementation details.

## Gas Performance

The following are a summary of the gas tests added in `/test/minter-suite-minters/RAM/MinterRAM/gas.test.ts`, which are intended to represent typical operations performed on the minter.

>note: all gas performance scales infinitely with project size

| Operation | Avg Gas Used | Comment |
| --- | --- | --- |
| Initial bids | 85k | first bid uses ~175k gas due to initialization of pointers |
| bids after max bids reached | 107k | first bid uses ~150k due to initialization of pointers |
| gas per admin/artist minted token | ~53k | assumes bulk-minting chunks of tokens at a time |

TODO:

- [x] Add tests for full coverage of new libraries and contracts

### CODE REVIEW

**Comment by ryley-o** [comment]
Commit for security audit: [d4e3fe7](https://github.com/ArtBlocks/artblocks-contracts/pull/1319/commits/d4e3fe7d73f977ed971b3f5d135b686c47579ae4)

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 1]
```diff
@@ -0,0 +1,2923 @@
...
+            // require project minter state C (Post-Auction, admin-artist-only,
+            //  not all bids handled)
+            ProjectMinterStates projectMinterState = getProjectMinterState({
+                projectId: projectId,
+                coreContract: coreContract
+            });
+            require(
+                projectMinterState ==
+                    ProjectMinterStates.PostAuctionSellOutAdminArtistMint,
+                "Only post-auction admin-artist-only"
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:1164]: nit: make <32 chars for contract size efficiency

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 2]
```diff
@@ -0,0 +1,2923 @@
...
+
+        // CHECKS
+        // require project minter state E (Post-Auction, all bids handled)
+        ProjectMinterStates projectMinterState = getProjectMinterState({
+            projectId: projectId,
+            coreContract: coreContract
+        });
+        require(
+            projectMinterState == ProjectMinterStates.PostAuctionAllBidsHandled,
+            "Only post-auction all bids handled"
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:1362]: nit: make <32 chars for contract size efficiency

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 3]
```diff
@@ -0,0 +1,2923 @@
...
+            // require project minter state D, or E (Post-Auction)
+            ProjectMinterStates projectMinterState = getProjectMinterState({
+                projectId: projectId,
+                coreContract: coreContract
+            });
+            require(
+                projectMinterState == ProjectMinterStates.PostAuctionOpenMint ||
+                    projectMinterState ==
+                    ProjectMinterStates.PostAuctionAllBidsHandled,
+                "Only post-auction open mint or all bids handled"
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:1437]: nit: make <32 chars for contract size efficiency

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 4]
```diff
@@ -0,0 +1,2923 @@
...
+     */
+    function _mintAndSettle(
+        uint256 projectId,
+        address coreContract,
+        uint256 projectPrice,
+        uint256 slotIndex,
+        uint256 bidId,
+        IMinterFilterV1 minterFilter,
+        uint256 minterRefundGasLimit
+    ) internal {
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:987]: nit: this should be private + follow order of functions

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 5]
```diff
@@ -0,0 +1,2923 @@
...
+     * @param minterRefundGasLimit Gas limit to use when refunding bidder
+     */
+    function _settleAndRefundBid(
+        uint256 projectId,
+        address coreContract,
+        uint256 projectPrice,
+        uint256 slotIndex,
+        uint256 bidId,
+        uint256 minterRefundGasLimit
+    ) internal {
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:1283]: should be private + follow order of functions

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 6]
```diff
@@ -0,0 +1,2923 @@
...
+        // load project config
+        RAMProjectConfig storage RAMProjectConfig_ = getRAMProjectConfig({
+            projectId: projectId,
+            coreContract: coreContract
+        });
+        // load bid
+        Bid storage bid = RAMProjectConfig_.bids[bidId];
+        // @dev bidderAddress previously checked not null
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:1291]: nit: this should be added to function natspec as an assumption

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 7]
```diff
@@ -0,0 +1,2923 @@
...
+        uint256 bidId,
+        IMinterFilterV1 minterFilter,
+        uint256 minterRefundGasLimit
+    ) internal {
+        // load project config
+        RAMProjectConfig storage RAMProjectConfig_ = getRAMProjectConfig({
+            projectId: projectId,
+            coreContract: coreContract
+        });
+        Bid storage bid = RAMProjectConfig_.bids[bidId];
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:993]: should this list an assumption in the natspec of the function that a non-null bidder address has already been checked?

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 8]
```diff
@@ -2574,15 +2556,18 @@ library RAMLib {
...
-            // preventing bids from being removed entirely from the last slot,
-            // which is acceptable and non-impacting for this minter
-            // @dev sets minBidSlotIndex to 511 if no more active bids, which
-            // is desired behavior for this minter
-            (RAMProjectConfig_.minBidSlotIndex, ) = _getMinSlotWithBid({
-                RAMProjectConfig_: RAMProjectConfig_,
-                startSlotIndex: slotIndex + 1
-            });
+            // update the minBidSlotIndex if it is the slotIndex of the ejected bid
+            if (RAMProjectConfig_.minBidSlotIndex == slotIndex) {
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2578]: yikes, good catch - I think this is a major bug fix, right?

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 9]
```diff
@@ -0,0 +1,2923 @@
...
+            projectId: projectId,
+            coreContract: coreContract
+        });
+        uint88 basePrice = RAMProjectConfig_.basePrice;
+        uint256 startBidValue = slotIndexToBidValue({
+            basePrice: basePrice,
+            slotIndex: startSlotIndex
+        });
+        // start search at next slot, incremented in while loop
+        uint256 currentSlotIndex = startSlotIndex;
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2837]: nit: should this just be a uint16?

**Review by ryley-o — COMMENTED** [review]
comments for changes up through [d65b164](https://github.com/ArtBlocks/artblocks-contracts/pull/1319/commits/d65b164ac37c0c1d3af67218428edd0d45701a6f)

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol [hunk 10]
```diff
@@ -162,29 +162,24 @@ library MaxInvocationsLib {
...
-                tokenInvocation <= localMaxInvocations,
-                "Max invocations reached"
-            );
-            // in typical case, update the local maxHasBeenInvoked value
-            // to true if the token invocation == minter local max invocations
-            // (enables gas efficient reverts after sellout)
-            if (tokenInvocation == localMaxInvocations) {
-                maxInvocationsProjectConfig.maxHasBeenInvoked = true;
-            }
+        // limit of 1e6 invocations per project.
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol:165]: This was flagged in the audit under I-03 - I think the comment may have gotten unlinked from the response here: https://docs.google.com/document/d/137_mzJURmzBH3mQZ7gN6O_Hb0w01LE0ocHISnWe8ka0/edit?disco=AAABJMYH-PM

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 9 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: Good call! Updated

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 7 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:993]: Added! The check in `adminArtistAutoMintTokensToWinners` isn't as explicit as the one in `directMintTokensToWinners`, it's checking that there is a slot with bid returned by `_getMaxSlotWithBid` but not explicitly checking the bidder address is not null.. Do you think it's redundant to check if the bidder address is not null before calling `_mintAndSettle` in that function?

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 6 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: Updated

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 5 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: Updated!

File: packages/contracts/contracts/libs/v0.8.x/ABHelpers.sol [hunk 11]
```diff
@@ -21,10 +21,8 @@ library ABHelpers {
         uint256 tokenId
     ) internal pure returns (uint256) {
         // int division properly rounds down
-        // @dev unchecked because will never divide by zero
-        unchecked {
-            return tokenId / ONE_MILLION;
-        }
+        // @dev no way to disable division by zero check in solidity, so not unchecked
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/ABHelpers.sol]: Updated the comment to include the compiler version

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 12]
```diff
@@ -0,0 +1,2923 @@
...
+            getProjectMinterState({
+                projectId: projectId,
+                coreContract: coreContract
+            }) == ProjectMinterStates.PreAuction,
+            "Only pre-auction"
+        );
+        // require base price >= 0.05 ETH
+        require(basePrice >= 0.05 ether, "Only base price gte 0.05 ETH");
+        // only future start time
+        // @dev minter already checks via min auction length check
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: Good catch! removed this

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 13]
```diff
@@ -0,0 +1,2923 @@
...
+     * Reverts if slotIndex > 511
+     * @param slotIndex Index of slot to set (between 0 and 511)
+     * @param RAMProjectConfig_ RAMProjectConfig to update
+     */
+    function _unsetBitmapSlot(
+        RAMProjectConfig storage RAMProjectConfig_,
+        uint256 slotIndex
+    ) private {
+        // revert if slotIndex >= NUM_SLOTS, since this is an invalid input
+        require(slotIndex < NUM_SLOTS, "Only slot index lt NUM_SLOTS");
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2631]: Updated

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 14]
```diff
@@ -0,0 +1,2923 @@
...
+     * Reverts if slotIndex > 511
+     * @param slotIndex Index of slot to set (between 0 and 511)
+     * @param RAMProjectConfig_ RAMProjectConfig to update
+     */
+    function _setBitmapSlot(
+        RAMProjectConfig storage RAMProjectConfig_,
+        uint256 slotIndex
+    ) private {
+        // revert if slotIndex >= NUM_SLOTS, since this is an invalid input
+        require(slotIndex < NUM_SLOTS, "Only slot index lt NUM_SLOTS");
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2602]: Updated

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol [hunk 15]
```diff
@@ -250,27 +250,11 @@ library MaxInvocationsLib {
         uint256 projectId,
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol:245]: should this be named `invocationsRemainOnCore`?

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 8 above)
Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2579]: yeah, huge bug fix nice 😓

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol [hunk 16]
```diff
@@ -351,34 +335,11 @@ library MaxInvocationsLib {
...
         address coreContract
     ) internal view returns (bool) {
-        // get max invocations from core contract
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol:355]: as discussed - this changes functionality so let's revert the change for this function 👍

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol (hunk 15 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol:245]: Good call, `invocationsRemainOnCore` is clearer.. Updated this + the call in SEALib

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol (hunk 16 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol]: Roger that! Updated

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 17]
```diff
@@ -0,0 +1,2926 @@
...
+            ) {
+                continue;
+            }
+            address bidderAddress = bid.bidder;
+            // require sender is bidder if requireSenderIsBidder is true
+            if (requireSenderIsBidder) {
+                require(msg.sender == bidderAddress, "Only sender is bidder");
+            } else {
+                // require bid exists
+                require(bidderAddress != address(0), "invalid Bid ID");
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:808]: I think this would be better to put before the `continue` operation, because I think reverting when an invalid Bid ID is supplied is a better UX than just skipping, and would help a mislead admin properly configure their tx. It comes at a slight gas increase, but probably worth it imo. thoughts?

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol (hunk 16 above)
Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol]: actually, I think I was just confused and this was actually just fine as it was 💀

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol [hunk 18]
```diff
@@ -352,6 +330,26 @@ library MaxInvocationsLib {
...
+    /**
+     * @notice Function returns the number of invocations available for a given
+     * project. Function checks the core contract's invocations and minter's max
+     * invocations, ensuring that the most limiting value is used, even if the
+     * local minter max invocations is stale.
+     * @param projectId The id of the project.
+     * @param coreContract The address of the core contract.
+     * @return Number of invocations available for the project.
+     */
+    function getInvocationsAvailable(
```
[the remark this answers is not in the document]
Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol:371]: actually, nvm

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 7 above)
Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:993]: yeah, I think the dev comment in `function adminArtistAutoMintTokensToWinners` that says `// @dev scrolling logic in State C ensures bid is not yet minted` is sufficient to not need a check, so this lgtm! 🙏

Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:993]: actually, maybe that comment should say: `// @dev scrolling logic in State C ensures bid **exists** is not yet minted`?

File: packages/contracts/contracts/libs/v0.8.x/BitMap.sol [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * @dev Library for using uint256 as a mapping to 256 bool values via a bit map.
+ * This is useful for storing a large number of bool values in a compact way.
+ * @dev This implementation is similar to OpenZeppelin's BitMaps library, but a
+ * single uint256 is used directly in memory instead of operating within a
+ * a mapping within a storage struct.
+ * This design limits the number of indices to 256, but is more gas efficient
+ * for use cases that fit within that limit. This is especially true for
+ * operations that require many reads/writes, since SLOAD/STTORE can be managed
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/BitMap.sol:14]:
```suggestion
 * operations that require many reads/writes, since SLOAD/SSTORE can be managed
```

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 20]
```diff
@@ -0,0 +1,2926 @@
...
+     * Reverts if slotIndex > 511
+     * @param slotIndex Index of slot to set (between 0 and 511)
+     * @param RAMProjectConfig_ RAMProjectConfig to update
+     */
+    function _setBitmapSlot(
+        RAMProjectConfig storage RAMProjectConfig_,
+        uint256 slotIndex
+    ) private {
+        // revert if slotIndex >= NUM_SLOTS, since this is an invalid input
+        // @dev no coverage as slot index out of range checked in placeBid
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2601]:
```suggestion
        // @dev no coverage as slot index out of range checked in placeBid and implicitly in topUpBid
```

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 21]
```diff
@@ -0,0 +1,2926 @@
...
+     * Reverts if slotIndex > 511
+     * @param slotIndex Index of slot to set (between 0 and 511)
+     * @param RAMProjectConfig_ RAMProjectConfig to update
+     */
+    function _unsetBitmapSlot(
+        RAMProjectConfig storage RAMProjectConfig_,
+        uint256 slotIndex
+    ) private {
+        // revert if slotIndex >= NUM_SLOTS, since this is an invalid input
+        // @dev no coverage as slot index out of range checked in placeBid
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2630]:
```suggestion
        // @dev no coverage as slot index out of range checked in placeBid and implicitly in topUpBid
```

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 22]
```diff
@@ -0,0 +1,2926 @@
...
+                    .slotsBitmapA
+                    .maxBitSet(
+                        // start at the end of the first bitmap
+                        uint8(255)
+                    );
+            }
+        }
+        // populate return value
+        if (!foundSlotWithBid) {
+            return uint16(NUM_SLOTS);
```
Comment (ryley-o) [inline packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2803]: missing if branch coverage here - should either add test for coverage or mark as unreachable as used

**Review by ryley-o — COMMENTED** [review]
✅  Consider this my approval (I can't approve since I opened this PR)
just minor coverage nits is all 🍾 really excited about this!

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 7 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:993]: Good call - that's better, updated!

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 23]
```diff
@@ -0,0 +1,2926 @@
...
+                    currentLatestMintedBidSlotIndex = _getMaxSlotWithBid({
+                        RAMProjectConfig_: RAMProjectConfig_,
+                        startSlotIndex: uint16(
+                            currentLatestMintedBidSlotIndex - 1
+                        )
+                    });
+                    require(
+                        currentLatestMintedBidSlotIndex < NUM_SLOTS,
+                        "slot with bid not found"
+                    );
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:931]: Added the comment!

File: packages/contracts/contracts/minter-suite/Minters/MinterRAMV0.sol [hunk 24]
```diff
@@ -0,0 +1,1205 @@
...
+     * @notice Purchases token for project `projectId` on core contract
+     * `coreContract` for auction that has ended, but not yet been sold out.
+     * @param projectId Project ID to purchase token for.
+     * @param coreContract Core contract address for the given project.
+     * @return tokenId Token ID of minted token
+     */
+    function purchase(
+        uint256 projectId,
+        address coreContract
+    ) external payable nonReentrant returns (uint256 tokenId) {
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/minter-suite/Minters/MinterRAMV0.sol:488]: Sounds great! Added a test for this here if you wouldn't mind TAL: https://github.com/ArtBlocks/artblocks-contracts/pull/1319/commits/94e8a5164e1fc8c46b7780b0691b72e6101fe3e6

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 22 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2804]: Added a dev comment here 🙏

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 25]
```diff
@@ -0,0 +1,2926 @@
...
+            _unsetBitmapSlot({
+                RAMProjectConfig_: RAMProjectConfig_,
+                slotIndex: slotIndex
+            });
+            // @dev reverts if removedSlotIndex was the maximum slot 511,
+            // preventing bids from being removed entirely from the last slot,
+            // which is acceptable and non-impacting for this minter
+            // @dev sets minBidSlotIndex to 512 if no more active bids, which
+            // is desired behavior for this minter
+            if (RAMProjectConfig_.minBidSlotIndex == slotIndex) {
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol:2579]: Good catch :) Added!

File: packages/contracts/contracts/minter-suite/Minters/MinterRAMV0.sol (hunk 24 above)
[the remark this answers is not in the document]
Comment (ryley-o) [reply packages/contracts/contracts/minter-suite/Minters/MinterRAMV0.sol:488]: noticed the coverage looked a little funny and found a quirk here: https://github.com/ArtBlocks/artblocks-contracts/commit/94e8a5164e1fc8c46b7780b0691b72e6101fe3e6#r140345875

**Review by lyaunzbe — APPROVED** [review]

File: packages/contracts/test/minter-suite-minters/constants.ts [hunk 26]
```diff
@@ -93,4 +93,45 @@ export const revertMessages = {
...
+    "Only post-auction open mint or all bids handled",
+  onlyStatePostAuctionBidsHandled: "Only post-auction all bids handled",
+  onlyStateErrorE1: "Only in state E1",
+  onlyEmergencyLTMax: "Only emergency hours lt max",
+  notInExtraTime: "Not allowed in extra time",
+  auctionTooShortRAM: "Auction too short",
+  onlyGTE0p05ETH: "Only base price gte 0.05 ETH",
+  onlyAdminArtistMintPeriod: "Only admin-artist mint period",
+  onlyNoAdminArtistMintPeriod: "Only no admin-artist mint period",
+  noPerviousAdminExtension: "No previous admin extension",
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/test/minter-suite-minters/constants.ts]: thanks! updated 🙏

**Review by lindsgil — APPROVED** [review]

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol [hunk 27]
```diff
@@ -0,0 +1,2923 @@
...
+            ProjectMinterStates projectMinterState = getProjectMinterState({
+                projectId: projectId,
+                coreContract: coreContract
+            });
+            require(
+                projectMinterState ==
+                    ProjectMinterStates.PostAuctionSellOutAdminArtistMint ||
+                    projectMinterState ==
+                    ProjectMinterStates.PostAuctionOpenMint,
+                "Only post-auction open mint or admin artist mint period"
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: nice! reverted this back to "Only State ..." messaging

File: packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol (hunk 1 above)
Comment (lindsgil) [reply packages/contracts/contracts/libs/v0.8.x/minter-libs/RAMLib.sol]: 🙏
